### ISSUE TITLE
trying to explain deconvlution kernel size in contextual attention

### ISSUE THREAD

**Issue by 12ycli** [JiahuiYu/generative_inpainting#294]
I have find some issues about this in [#73](https://github.com/JiahuiYu/generative_inpainting/issues/73) and [#22](https://github.com/JiahuiYu/generative_inpainting/issues/22) and according to the links provided by @JiahuiYu , I did some experiments and tried to explain it here. This also confuses me for several days and I hope this helps.
At first, I generated a specified image to do the experiment. The first channel of this image is 4*6, and the pixel value is like below, which helps us for confirmation.
```
 0  1  2  3  4  5
 6  7  8  9 10 11
... 
```
```
def gen_simple_image():
    height,width,channel = 4,6,3
    img = np.reshape(np.arange(height*width*channel),(channel,height,width))
    img = np.transpose(img,[1,2,0])
    cv2.imwrite('simple_image.bmp',img)
```
The two images below corresponds to the situation in one dimension. We can see in this situation, kernel size is always two times of stride. This explains why the code below has number 2. 
https://github.com/JiahuiYu/generative_inpainting/blob/06cd62cfca8c10c349b451fa33d9cbb786bfaa20/inpaint_ops.py#L242
![2019-07-16_224336](https://user-images.githubusercontent.com/10368281/61304242-74487600-a81b-11e9-8083-164a9f1510dd.png)
![2019-07-16_225146](https://user-images.githubusercontent.com/10368281/61304820-56c7dc00-a81c-11e9-9790-c2ac8cd5bf06.png)
Moreover, we can see that two values cover the same pixel so we want to average them by 2. Our situation is two dimension and we can image we should average them by 4 in the code below.
https://github.com/JiahuiYu/generative_inpainting/blob/06cd62cfca8c10c349b451fa33d9cbb786bfaa20/inpaint_ops.py#L307
We can do the experiment and check whether the output y in function `contextual_attention` is the same as its input f or b(f and b are the same). It should be very closed from the intention of `contextual_attention`——we can borrow any patches from b and paste to y according to the similarity between b and f.(Some may wonder that, in proposed model, the output y will always the same b. Actually, it won't if the mask is all zero, which is the situation in our experiment. The mask stops y from borrowing patches in b in masked region.) Result is shown below, we can see that except some pixels in the boundary, other pixels are exactly the same.
![2019-07-16_230429](https://user-images.githubusercontent.com/10368281/61305858-27b26a00-a81e-11e9-80a0-d4140da738da.png)
Should convolution kernel size(w) the same as deconvolution kernel size(raw_w) as stated in [#22](https://github.com/JiahuiYu/generative_inpainting/issues/22). In my opinion, convolution and its subsequent softmax decide which the deconvolution kernel should be put where. Thus the kernel size of deconvolution do not matter much. We can try to change parameter ksize from 3 to 5. y should be the same as above. Actually it is as I did the experiment.
I wondered if we can change 2 or/and 4 in the code above to get the same effect. I found another pattern shown below.
![2019-07-16_234235](https://user-images.githubusercontent.com/10368281/61308755-7b738200-a823-11e9-8c23-a15e0889b65a.png)
![2019-07-16_234329](https://user-images.githubusercontent.com/10368281/61308788-8e865200-a823-11e9-90db-f0683c3a0ee8.png)
According to the pattern shown above, I change 2 to 1 and change 4 to 1 and code shown below. Finally I got y exactly the same as input b. 
```
    raw_w = tf.extract_image_patches(
        b, [1,kernel,kernel,1], [1,kernel,kernel,1], [1,1,1,1], padding='SAME')
```
![2019-07-16_234754](https://user-images.githubusercontent.com/10368281/61309106-2b48ef80-a824-11e9-829d-36fe01c5b215.png)
Besides, in [#22](https://github.com/JiahuiYu/generative_inpainting/issues/22), @JiahuiYu said 'Basically larger patch mean more accurate of original resolution.'. In my opinion, convolution result gives rough result in low resolution. Large patch covers more pixels and every pixel gets more to average in deconvolution, which may gives more accurate result. Below image shows larger patches.
![2019-07-17_000538](https://user-images.githubusercontent.com/10368281/61310337-af9c7200-a826-11e9-932d-a8057ebe4e0e.png)
Above discussion is based on the effect of deconvolution is to paste its kernels all around, which is decided by convolution and softmax output yi. I can comprehend this from the perspective that deconvolution is another kind of convolution as stated in [1]. However I can't explain it in more details. May someone can do that for us.

Reference:
[1] A guide to convolution arithmetic for deep learning

**Comment by 12ycli** [comment]
@JiahuiYu I wonder if we really have to use deconvolution to copy patches of b to y.
I wrote a toy function to do this.
```
def deconv():
    img = np.zeros(shape=(1,5,5,1))
    img[0,3,3,0] = 1
    print('sigmoid output:\n', img[0,:,:,0])
    filter = np.reshape(np.arange(9),(3,3))
    print('filter:\n', filter)
    filter = np.fliplr(np.flipud(filter))
    # print('filter:\n',filter)
    filter = np.expand_dims(filter,2)
    filter = np.expand_dims(filter,3)

    output = tf.nn.conv2d(img,filter,strides=[1,1,1,1],padding='SAME')

    tf.InteractiveSession()
    print('output:\n', output.eval()[0,:,:,0])
```

sigmoid output:
 [[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0.]]
filter:
 [[0 1 2]
 [3 4 5]
 [6 7 8]]
output:
 [[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 2.]
 [0. 0. 3. 4. 5.]
 [0. 0. 6. 7. 8.]]

**Comment by 12ycli** [comment]
However, toy function above have some transpose and convolution operation. Maybe this is deconvolution? Deconvolution have uneven overlaps because it have to keep the resolution right? I will hold this in my head and figure it out someday.

**Comment by 12ycli** [comment]
Yes, toy function above is deconvolution.
```
def cheak_deconv():
    image = np.reshape(np.arange(25.0),(5,5))

    image = np.expand_dims(image,0)
    image = np.expand_dims(image,3)

    filter = np.reshape(np.arange(9.0),(3,3))
    transposed_filter = np.fliplr(np.flipud(filter))
    filter = np.expand_dims(filter,2)
    filter = np.expand_dims(filter,3)
    transposed_filter = np.expand_dims(transposed_filter,2)
    transposed_filter = np.expand_dims(transposed_filter,3)

    conv = tf.nn.conv2d(image,filter,strides=[1,1,1,1],padding='VALID')

    tf.InteractiveSession()

    transposed_conv = tf.nn.conv2d_transpose(conv,filter,[1,5,5,1],padding='VALID')
    print('tensorflow conv2d_transpose:\n', transposed_conv.eval()[0,:,:,0])

    padded_conv_output = tf.pad(conv,[[0,0],[2,2],[2,2],[0,0]])
    output3 = tf.nn.conv2d(padded_conv_output, transposed_filter, strides=[1,1,1,1], padding='VALID')
    print('simulate deconv using transposed filter and conv:\n', output3.eval()[0,:,:,0])
```
tensorflow conv2d_transpose:
 [[    0.   312.   972.  1080.   768.]
 [  936.  2784.  5616.  4896.  3048.]
 [ 3348.  8496. 15552. 12528.  7380.]
 [ 4968. 11424. 19440. 14688.  8232.]
 [ 4032.  8952. 14796. 10872.  5952.]]
simulate deconv using transposed filter and conv:
 [[    0.   312.   972.  1080.   768.]
 [  936.  2784.  5616.  4896.  3048.]
 [ 3348.  8496. 15552. 12528.  7380.]
 [ 4968. 11424. 19440. 14688.  8232.]
 [ 4032.  8952. 14796. 10872.  5952.]]

**Comment by JiahuiYu** [comment]
@12ycli Hi, thanks for your interest in our work first.

And thanks for your continuous efforts on explaining these stuff to others. I will keep this thread open so others can also have a references. If you can slightly reorganize of your contents in this thread, I would be super appreciated. Thanks!
